src/machine.rs

```diff
@@ -54,15 +54,18 @@ pub struct FrameData<'tcx> {
     /// for the start of this frame. When we finish executing this frame,
     /// we use this to register a completed event with `measureme`.
     pub timing: Option<measureme::DetachedTiming>,
+
+    pub protected_tags: Vec<SbTag>,
```

Comment:
Can you put the two Stacked Borrows things into a single type exported by Stacked Borrows?
We should have at most one field per "subsystem" here.

Also please add a doc comment explaining this. In particular, why is it a `Vec` and not a `FxHashSet`?

Reply:
Oh I see, you just iterate over this to clean up the global `protected_tags`. That's some annoying bookkeeping...

What happens with the following alternatives: we have a `protected_tags: FxHashSet<SbTag>` for each frame. Checking if a tag's protector is still active iterates over each stack frame and checks all sets for membership. That's more expensive to check but we avoid the memory and time for the cleanup bookkeeping.

Reply:
I just tried implementing this, and it looks like this approach is significantly slower for some workloads, but generally mixed across the benchmark suite. I think the penalty is because Swiss Tables are organized into groups, which I think are 8 elements each. Since the lookups are done with SIMD, the overhead in CPU and memory is always close to the nearest multiple of 8.

It looks like about half of the sets contain 1 element and 98% contain 1 or 2. I think we might be able to salvage the runtime by using a `Vec`... but that might be strange to read.

Reply:
Ah good point, the number of protectors is basically equal to the number of function arguments of reference type. (With `-Zmiri-retag-fields`, also counting references inside ADTs.) So, it's going to be small. Which probably makes `Vec` a reasonable choice? Could make it a sorted `Vec` and use binary search, though that makes the access pattern more random so not sure if it is a good idea.

Reply:
We can print a backtrace without panicking, can't we?
I think benchmarks should have exit code 0.

Reply:
(I know you already know this, but for continuity) I finally figured out how to extract the benchmark I want: https://github.com/rust-lang/miri/pull/2352

With the array bumped back up to 4096 bytes, I get this benchmark runtime before this PR:
```
Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/slice-get-unchecked/Cargo.toml
  Time (mean ± σ):      6.646 s ±  0.098 s    [User: 5.666 s, System: 0.949 s]
  Range (min … max):    6.539 s …  6.768 s    5 runs
```
With the current state of this PR:
```
Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/slice-get-unchecked/Cargo.toml
  Time (mean ± σ):      4.835 s ±  0.055 s    [User: 4.407 s, System: 0.363 s]
  Range (min … max):    4.772 s …  4.909 s    5 runs
```
And if I remove the global HashSet and instead walk every stack frame to see if an `Item` is still protected:
```
Benchmark 1: cargo +miri miri run --manifest-path bench-cargo-miri/slice-get-unchecked/Cargo.toml
  Time (mean ± σ):      5.304 s ±  0.090 s    [User: 4.857 s, System: 0.354 s]
  Range (min … max):    5.183 s …  5.393 s    5 runs
```
So that's a 10% regression.

I grabbed some stats from `println!` + `counts` and I think I can explain what's going on here.
Here is the distribution of the size of the global HashSet when we go to remove protectors from it:
```
83332 counts
(  1)    20555 (24.7%, 24.7%): 18
(  2)    12372 (14.8%, 39.5%): 14
(  3)    12367 (14.8%, 54.4%): 17
(  4)    12343 (14.8%, 69.2%): 16
(  5)    12313 (14.8%, 83.9%): 20
(  6)     8275 ( 9.9%, 93.9%): 19
(  7)     4111 ( 4.9%, 98.8%): 21
(  8)       96 ( 0.1%, 98.9%): 13
(  9)       89 ( 0.1%, 99.0%): 15
( 10)       88 ( 0.1%, 99.1%): 10
```
Here is the distribution of the size of the global HashSet when we are popping a protected item:
```
17055803 counts
(  1) 16806170 (98.5%, 98.5%): 15
(  2)    61658 ( 0.4%, 98.9%): 18
(  3)    37114 ( 0.2%, 99.1%): 17
(  4)    37103 ( 0.2%, 99.3%): 14
(  5)    36964 ( 0.2%, 99.5%): 20
(  6)    32922 ( 0.2%, 99.7%): 16
(  7)    24822 ( 0.1%, 99.9%): 19
(  8)    16432 ( 0.1%,100.0%): 21
(  9)      304 ( 0.0%,100.0%): 13
( 10)      249 ( 0.0%,100.0%): 12
```
And here is the distribution of the number of stack frames in the program when we are popping a protected item:
```
17055803 counts
(  1) 16789696 (98.4%, 98.4%): 19
(  2)    82149 ( 0.5%, 98.9%): 16
(  3)    57573 ( 0.3%, 99.3%): 17
(  4)    49400 ( 0.3%, 99.5%): 18
(  5)    37198 ( 0.2%, 99.8%): 14
(  6)    20783 ( 0.1%, 99.9%): 15
(  7)    12387 ( 0.1%,100.0%): 20
(  8)     4154 ( 0.0%,100.0%): 21
(  9)      397 ( 0.0%,100.0%): 12
( 10)      373 ( 0.0%,100.0%): 13
```

The total number of counts at the top of each distribution is itself revealing, checking for a protector happens ~1,000 times more than returning from a stack frame. I think that's why updating a global HashSet upon function return hardly shows up in a profile. I think the ratio is so dramatic because in the slice-get-unchecked benchmark, each function call which protects the slice generates thousands of protected `Item`s per call.

Then, if we attempt to walk every stack frame to figure out if the protector is still active, we are iterating over about as many stack frames as we have total protected tags. On top of that, we need to search each frame (by searching a SmallVec or HashSet) for the tag that's being popped.

So overall, I'm not sure this makes a particularly strong argument for either implementation. I think it makes a pretty solid argument against memory overhead as a reason to avoid the global HashSet, the global set just never gets that big.

In larger programs, I would expect to see deeper call stacks. So that probably means more total protectors at any given time, but also if the call stacks are deeper while we are popping lots of items this could penalize either implementation.

In concurrent programs I would expect a bigger penalty for the "iterate over every stack frame" implementation, because we will be multiplying the amount of linear searching we have to do, but the global HashSet maintains all its nice constant-time properties.

Overall this is primarily guesswork, and I'm not sure a 10% improvement on printing a backtrace from a failed test is worth all this sweat. Whichever implementation you prefer after reading this, I think we just generally keep the other one in mind and if someone comes up with an example program that blows up the protector checking runtime, we consider changing it.

Reply:
Thanks for the detailed analysis!

Are there any perf differences between the two options for the other benchmarks?

> And if I remove the global HashSet and instead walk every stack frame to see if an Item is still protected:

Just for clarity: which version of the per-frame "protected" set was this used -- a HashSet, a Vec, or a SmallVec? (I am not asking for you to also try the other variants, just want to know what was being measured.)

> checking for a protector happens ~1,000 times more than returning from a stack frame

That is by far the best argument to me, so please put it in the code as justification for optimizing the protector-check case to be a single hashset lookup rather than a full stack walk.

Reply:
> Are there any perf differences between the two options for the other benchmarks?

0-5%. The global HashSet is always faster or within noise.

> which version of the per-frame "protected" set was this used -- a HashSet, a Vec, or a SmallVec?

`SmallVec<[SbTag; 2]>`